### DISCUSSION TITLE
Escape column names that conflict with Rust keywords in CLI

### DISCUSSION THREAD

**Discussion by tqwewe** [SeaQL/sea-orm#82] [Q&A] [answered]
When generating models with the sea-orm-cli, if you have a column with a name that conflicts with a Rust keyword, the cli crashes and is not handled correctly.

For example, an SQL table:
```
| column_name | data_type |
|-------------------------|
| id          | serial    |
| type        | varchar   |
```
I get an error:
```
11 | ... { pub id : i64 , pub type : Option < String > }
   |                          ^^^^ expected identifier, found keyword
```
This is caused because the column `type` conflicts with the Rust keyword `type`.

Any rust columns conflicting should be renamed to resolve this issue.
https://doc.rust-lang.org/reference/keywords.html

**Comment by tyt2y3** [comment]
Any suggestion on the naming convention? How about an underscore at the end?

**Comment by tqwewe** [comment]
I didn't make any suggestions for this because I am not sure 😄 

I guess `type_` with an underscore could work.. or `type_col` (adding `_col`). I'm not sure of any better options than these.

**Comment by billy1624** [comment, marked as answer]
I think we can leave it for the user to decide how to resolve the name conflict.

Personally think that neither `r#type` nor `type_` be a good option. For me, I would simply choose another name to avoid conflicting with Rust keywords.

```rust
#[derive(Clone, Debug, PartialEq, DeriveModel, DeriveActiveModel)]
pub struct Model {
    pub id: i32,
    pub type_of: String,
}

#[derive(Copy, Clone, Debug, EnumIter, DeriveCustomColumn)]
pub enum Column {
    Id,
    TypeOf,
}

impl IdenStatic for Column {
    fn as_str(&self) -> &str {
        match self {
            Self::TypeOf => "type",
            _ => self.default_as_str(),
        }
    }
}
```

**Reply by jsalort** [reply]
Is this answer still correct today ?
I am getting this error:
```
error[E0412]: cannot find type `Entity` in this scope
 --> src/entity/grandeurs.rs:5:35
  |
5 | #[derive(Clone, Debug, PartialEq, DeriveModel, DeriveActiveModel)]
  |                                   ^^^^^^^^^^^
  |
  = note: this error originates in the derive macro `DeriveModel` (in Nightly builds, run with -Z macro-backtrace for more info)
help: you might have meant to use the associated type
  |
5 | #[derive(Clone, Debug, PartialEq, Self::DeriveModel, DeriveActiveModel)]

```

**Comment by tqwewe** [comment]
The problem is that users will integrate existing databases with SeaORM.. and usually you don't want your framework / language to dictate your database, it should be flexible enough to avoid issues like this

**Comment by billy1624** [comment]
Good point!

**Comment by tyt2y3** [comment]
I think we didn't require the database schema to change here. Just a change in Rust's Entity definition. And the trick is to simply use a non Rust keyword in the enum variant and struct attribute, and map to the actual underlying table column.
I think it is a neat solution, and certainly looks better than _ or #
The only problem is whether the code generator should be smart enough to do this automatically.
I think it is a difficult task, only because the Rust keywords list is growing, and is case sensitive.

**Comment by Stumblinbear** [comment]
Has any progress been made on this? My schema uses `type` in a few places, and I'd rather not go through the effort to update this everywhere. Just `type_` would be my preferred solution.

**Reply by tyt2y3** [reply]
Sadly, not yet.
It is currently possible to rename the column mapping (as shown above), but a little clunkily.
The bigger problem is that the `DeriveColumn` macro (hence `DeriveEntityModel`) lacks a `rename` annotation, and thus it is only possible using the expanded format now.
We hope to add that soon.

**Reply by baoyachi** [reply]
Will `_type` consider？

**Reply by Stumblinbear** [reply]
Personally I think `type_` is a better solution, because then it won't be pulled out of alphabetical order in IDE autocomplete

**Reply by DanielHZhang** [reply]
What about using `ty` or `typ`? It's shorter, and personally I think it looks better than `type_`.
